File: Niveau 3/Niveau 3.11/galerie.py
#-------------------------------------------------------------------------------
# Name:        Galerie souterraine
# Purpose:     France-IOI niveau 3.11
#
# Author:      Sébastien
#
# Created:     21/10/2022
#-------------------------------------------------------------------------------

import sys
input = sys.stdin.readline

def main():
    h, l, d = map(int, input().split())
    plan = [[(c == 0) for c in map(int, input().split())] for _ in range(h)]

    x = 0
    y = 0
    n = 1

    while x != l-1 or y != h-1:
        if n % (d+1) == 0:
            print(y, x)
        plan[y][x] = False
        if   x > 0   and plan[y][x-1]: x -= 1
        elif x < l-1 and plan[y][x+1]: x += 1
        elif y > 0   and plan[y-1][x]: y -= 1
        else:                          y += 1
        n += 1
    if n % (d+1) == 0:
        print(y, x)

if __name__ == '__main__':
    main()

"""
9 11 7
0 0 1 0 0 0 1 1 1 1 1
1 0 0 0 1 0 1 0 0 0 0
1 1 1 1 1 0 1 0 1 1 0
1 1 0 0 0 0 1 0 0 1 0
1 1 0 1 1 1 1 1 0 1 0
1 1 0 1 1 1 1 0 0 1 0
1 1 0 0 0 0 1 0 1 0 0
1 1 1 1 1 0 1 0 1 0 1
1 1 1 1 1 0 0 0 1 0 0
---
2 2 0
0 0
1 0
"""

File: Niveau 3/Niveau 3.11/galerie.hs
--------------------------------------------------------------------------------
-- Name:        Galerie souterraine
-- Purpose:     France-IOI niveau 3.11
--
-- Author:      Sébastien
--
-- Created:     22/10/2022
--------------------------------------------------------------------------------

import Control.Monad
import Data.Array

type Coord = (Int, Int) -- (y: ligne, x: colonne)

main :: IO ()
main = do
    (h:l:d:_) <- fmap (map read . words) getLine :: IO [Int]
    plan <- fmap (listArray ((0,0), (h-1,l-1)) . map read . words) getContents :: IO (Array Coord Int)
    afficher plan (-1, -1) (0, 0) (h, l) 1 (d+1)
    
afficher :: Array Coord Int -- Plan
         -> Coord           -- La position précédente (pour tester si on revient en arrière)
         -> Coord           -- La position courante
         -> Coord           -- La taille du plan
         -> Int             -- Le nombre de cases visitées
         -> Int             -- La distance entre deux conduits d'aération
         -> IO ()
afficher plan prec position taille n d = do
    if n `mod` d == 0 
    then putStrLn $ show (fst position) ++ " " ++ show (snd position)
    else return ()
    case avancer plan prec position taille of
        Just a -> afficher plan position a taille (n+1) d
        Nothing -> return ()
    
avancer :: Array Coord Int -- Plan
        -> Coord           -- La position précédente (pour tester si on revient en arrière)
        -> Coord           -- La position courante
        -> Coord           -- La taille du tableau
        -> Maybe Coord     -- La position suivante
avancer plan prec (y, x) (h, l)
    | y == h-1 && x == l-1                              = Nothing
    | x < l-1 && plan!(y, x+1) == 0 && (y, x+1) /= prec = Just (y, x+1)
    | x > 0   && plan!(y, x-1) == 0 && (y, x-1) /= prec = Just (y, x-1)
    | y < h-1 && plan!(y+1, x) == 0 && (y+1, x) /= prec = Just (y+1, x)
    | otherwise                                         = Just (y-1, x)
